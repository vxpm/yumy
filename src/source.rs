use nonmax::NonMaxU32;
use owo_colors::Style;

/// Unit struct that represents the absence of
/// a source in a diagnostic.
#[derive(Debug, Clone, Copy)]
pub struct NoSource;

/// A span into the source of a [`crate::Diagnostic`].
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct SourceSpan {
    start: NonMaxU32,
    end: u32,
}

impl SourceSpan {
    /// Creates a new source span. `start` and `end`
    /// are byte indexes into the source.
    pub fn new(start: u32, end: u32) -> Self {
        assert!(end >= start);
        Self {
            start: NonMaxU32::new(start).expect("Start is non-max"),
            end,
        }
    }

    /// The start of this span. Inclusive.
    #[inline]
    pub fn start(&self) -> u32 {
        self.start.get()
    }

    /// The end of this span. Exclusive.
    #[inline]
    pub fn end(&self) -> u32 {
        self.end
    }

    /// The length of this span.
    #[inline]
    pub fn len(&self) -> u32 {
        self.end - self.start()
    }

    /// Does this span contain the given value?
    #[inline]
    pub fn contains(&self, value: u32) -> bool {
        (self.start() <= value) && (value < self.end)
    }
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub struct SourceLine<'src> {
    pub span: SourceSpan,
    pub line: &'src str,
}

impl<'src> SourceLine<'src> {
    pub fn new(line: &'src str, span: SourceSpan) -> Self {
        Self { span, line }
    }
}

/// A source of text to use with a diagnostic.
#[derive(Debug, Clone)]
pub struct Source<'src> {
    src: &'src str,
    name: Option<&'src str>,
    style: Option<Style>,
    lines: Vec<SourceLine<'src>>,
}

impl<'src> Source<'src> {
    fn lines_of(src: &str) -> Vec<SourceLine> {
        let base_addr = src.as_ptr();
        let lines = src.lines().map(|line| {
            let line_addr = line.as_ptr();
            let offset = (line_addr as usize)
                .checked_sub(base_addr as usize)
                .expect("line should always have higher address");
            let end = offset + line.len();

            SourceLine {
                span: SourceSpan::new(offset as u32, end as u32),
                line,
            }
        });

        lines.collect()
    }

    /// Creates a new source.
    pub fn new(src: &'src str, name: Option<&'src str>) -> Self {
        Self {
            src,
            name,
            style: None,
            lines: Self::lines_of(src),
        }
    }

    /// Creates a new source with the given style.
    pub fn styled(src: &'src str, name: Option<&'src str>, style: Style) -> Self {
        Self {
            src,
            name,
            style: Some(style),
            lines: Self::lines_of(src),
        }
    }

    pub fn src(&self) -> &'src str {
        self.src
    }

    pub fn name(&self) -> Option<&'src str> {
        self.name
    }

    pub fn style(&self) -> Option<Style> {
        self.style
    }

    pub fn line(&self, index: u32) -> Option<SourceLine<'src>> {
        self.lines.get(index as usize).copied()
    }

    pub fn lines(&self) -> impl Iterator<Item = &SourceLine> {
        self.lines.iter()
    }

    pub(crate) fn line_index_of_byte(&self, byte_index: u32) -> Option<u32> {
        if byte_index > self.src.len() as u32 {
            return None;
        }

        // TODO: change to a binary search
        for (index, line) in self.lines.iter().enumerate() {
            if line.span.contains(byte_index) {
                return Some(index as u32);
            }

            if line.span.start() > byte_index {
                return Some((index - 1) as u32);
            }
        }

        None
    }
}

#[cfg(test)]
mod test {
    use super::*;

    const SAMPLE: &str = include_str!("../sample.txt");

    #[test]
    fn test_lines() {
        let src = Source::new(SAMPLE, None);
        let mut lines = src.lines();

        assert_eq!(
            Some(&SourceLine {
                span: SourceSpan::new(0, 20),
                line: "hello there darling!"
            }),
            lines.next()
        );

        // notice how index 20 is not included in any line
        // since it is the \n character!

        assert_eq!(
            Some(&SourceLine {
                span: SourceSpan::new(21, 28),
                line: "this is"
            }),
            lines.next()
        );

        // the same goes for 28

        assert_eq!(
            Some(&SourceLine {
                span: SourceSpan::new(29, 29),
                line: ""
            }),
            lines.next()
        );

        // index 29 is the \n! but it is also not included since
        // the line is empty (29..29)

        assert_eq!(
            Some(&SourceLine {
                span: SourceSpan::new(30, 46),
                line: "a sample text :)"
            }),
            lines.next()
        );
    }
}
