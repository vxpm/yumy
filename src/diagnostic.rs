mod body;

/// Module for diagnostic configuration related items.
pub mod config;

use self::{body::BodyDescriptor, config::Config};
use super::source::{NoSource, Source, SourceSpan};
use owo_colors::{OwoColorize, Style};
use std::{
    io::{BufWriter, Write},
    ops::Range,
};

/// A label is a message that points to a specific
/// part of the source of a [`Diagnostic`].
#[derive(Debug, Clone)]
pub struct Label {
    /// The message of this label.
    pub message: String,
    /// The span this label refers to.
    pub span: SourceSpan,
    /// The indicator style of this label.
    pub indicator_style: Option<Style>,
}

impl Label {
    /// Creates a new label.
    pub fn new<S, M>(span: S, message: M) -> Self
    where
        S: Into<SourceSpan>,
        M: ToString,
    {
        Self {
            message: message.to_string(),
            span: span.into(),
            indicator_style: None,
        }
    }

    /// Creates a new label with the given style for it's indicator.
    pub fn styled<S, M>(span: S, message: M, style: Style) -> Self
    where
        S: Into<SourceSpan>,
        M: ToString,
    {
        Self {
            message: message.to_string(),
            span: span.into(),
            indicator_style: Some(style),
        }
    }

    /// Returns the line range of this label in the given source.
    ///
    /// # Panics
    /// Panics if the span is out of bounds.
    fn line_range(&self, src: &Source) -> Range<usize> {
        src.line_range_of_span(self.span)
            .expect("label should have span in range")
    }

    /// Returns whether this label is singleline or not.
    ///
    /// # Panics
    /// Panics if [`Label::line_range`] would panic.
    fn is_singleline(&self, src: &Source) -> bool {
        let line_range = self.line_range(src);
        line_range.start + 1 == line_range.end
    }
}

/// A diagnostic.
#[derive(Debug, Clone)]
pub struct Diagnostic<Src> {
    message: String,
    labels: Vec<Label>,
    footnotes: Vec<String>,
    source: Src,
}

impl Diagnostic<NoSource> {
    /// Create a new diagnostic without an associated source.
    #[inline]
    pub fn new<M>(message: M) -> Self
    where
        M: ToString,
    {
        Self {
            message: message.to_string(),
            labels: Vec::new(),
            footnotes: Vec::new(),
            source: NoSource,
        }
    }

    /// Attach a source to this diagnostic.
    #[inline(always)]
    pub fn with_source(self, source: Source<'_>) -> Diagnostic<Source<'_>> {
        Diagnostic {
            message: self.message,
            labels: self.labels,
            footnotes: self.footnotes,
            source,
        }
    }
}

impl<Src> Diagnostic<Src> {
    /// Add a [`Label`] to this diagnostic.
    #[inline(always)]
    pub fn with_message<M>(mut self, message: M) -> Self
    where
        M: ToString,
    {
        self.message = message.to_string();
        self
    }

    /// Add a [`Label`] to this diagnostic.
    #[inline(always)]
    pub fn add_label(&mut self, label: Label) {
        self.labels.push(label);
    }

    /// Add a [`Label`] to this diagnostic.
    #[inline(always)]
    pub fn with_label(mut self, label: Label) -> Self {
        self.add_label(label);
        self
    }

    /// Replaces the [`Label`]s of this diagnostic.
    #[inline(always)]
    pub fn with_labels(mut self, labels: Vec<Label>) -> Self {
        self.labels = labels;
        self
    }

    /// Add a footnote to this diagnostic. A footnote is a message
    /// shown after the body of a diagnostic.
    #[inline(always)]
    pub fn add_footnote<F>(&mut self, footnote: F)
    where
        F: ToString,
    {
        self.footnotes.push(footnote.to_string());
    }

    /// Add a footnote to this diagnostic. A footnote is a message
    /// shown after the body of a diagnostic.
    #[inline(always)]
    pub fn with_footnote<F>(mut self, footnote: F) -> Self
    where
        F: ToString,
    {
        self.add_footnote(footnote.to_string());
        self
    }
}

impl<'src> Diagnostic<Source<'src>> {
    /// Writes the header of this diagnostic. The header is composed of:
    /// 01. The error message of the diagnostic (`self.message`).
    /// 02. The name of the source of the error (`self.source`).
    fn write_header<W>(
        &self,
        writer: &mut W,
        config: &Config,
        line_number_width: usize,
    ) -> std::io::Result<()>
    where
        W: Write,
    {
        // write the error message
        writeln!(writer, "{}", self.message)?;

        // write source
        write!(writer, "{:padding$}", "", padding = line_number_width)?;
        writeln!(
            writer,
            " {} {}{}{}",
            '@'.style(config.styles.left_column),
            '['.style(config.styles.left_column),
            self.source
                .name()
                .unwrap_or("unknown")
                .style(config.styles.source_name),
            ']'.style(config.styles.left_column)
        )?;

        Ok(())
    }

    fn write_header_compact<W>(&self, writer: &mut W, config: &Config) -> std::io::Result<()>
    where
        W: Write,
    {
        writeln!(writer, "{}", self.message)?;
        writeln!(
            writer,
            "{} {}{}{}",
            '@'.style(config.styles.left_column),
            '['.style(config.styles.left_column),
            self.source
                .name()
                .unwrap_or("unknown")
                .style(config.styles.source_name),
            "]:".style(config.styles.left_column)
        )?;
        Ok(())
    }

    /// Writes the body of this diagnostic. The body is composed of:
    /// 01. The source line's for which this diagnostic's labels refer to.
    /// 02. The labels themselves (`self.labels`).
    fn write_body<W>(
        &self,
        writer: &mut W,
        config: &Config,
        body_descriptor: BodyDescriptor,
    ) -> std::io::Result<()>
    where
        W: Write,
    {
        body_descriptor.write_to(writer, config)
    }

    fn write_body_compact<W>(&self, writer: &mut W, config: &Config) -> std::io::Result<()>
    where
        W: Write,
    {
        for label in &self.labels {
            let range = label.line_range(&self.source);
            if range.start + 1 == range.end {
                writeln!(
                    writer,
                    "{} {}{}{}{}{}",
                    config.charset.vertical_bar.style(config.styles.left_column),
                    '['.style(config.styles.left_column),
                    "line ".style(config.styles.source),
                    range.start.style(config.styles.source),
                    "]: ".style(config.styles.left_column),
                    label.message
                )?;
            } else {
                writeln!(
                    writer,
                    "{} {}{}{:?}{}{}",
                    config.charset.vertical_bar.style(config.styles.left_column),
                    '['.style(config.styles.left_column),
                    "lines ".style(config.styles.source),
                    range.style(config.styles.source),
                    "]: ".style(config.styles.left_column),
                    label.message
                )?;
            }
        }

        Ok(())
    }

    /// Writes the footnotes of this diagnostic (`self.footnotes`).
    fn write_footnotes<W>(
        &self,
        writer: &mut W,
        config: &Config,
        line_number_width: usize,
    ) -> std::io::Result<()>
    where
        W: Write,
    {
        for footnote in self.footnotes.iter() {
            write!(
                writer,
                "{:padding$} {} ",
                "",
                '>'.style(config.styles.footnote_indicator),
                padding = line_number_width
            )?;
            writeln!(writer, "{}", footnote)?;
        }

        Ok(())
    }

    fn write_footnotes_compact<W>(&self, writer: &mut W, config: &Config) -> std::io::Result<()>
    where
        W: Write,
    {
        for footnote in &self.footnotes {
            write!(writer, "{} ", '>'.style(config.styles.footnote_indicator))?;
            writeln!(writer, "{}", footnote)?;
        }

        Ok(())
    }

    /// Writes this diagnostic to the given [`Write`]r using the specified [`Config`].
    pub fn write_to<W>(&self, writer: &mut W, config: &Config) -> std::io::Result<()>
    where
        W: Write,
    {
        let body_descriptor = body::BodyDescriptor::new(self.source.clone(), self.labels.clone());
        let line_number_width = body_descriptor.line_number_width;

        self.write_header(writer, config, line_number_width)?;
        self.write_body(writer, config, body_descriptor)?;
        self.write_footnotes(writer, config, line_number_width)?;

        writeln!(writer)?;
        Ok(())
    }

    /// Writes this diagnostic to the given [`Write`]r using the specified [`Config`]
    /// in compact mode.
    pub fn write_to_compact<W>(&self, writer: &mut W, config: &Config) -> std::io::Result<()>
    where
        W: Write,
    {
        self.write_header_compact(writer, config)?;
        self.write_body_compact(writer, config)?;
        self.write_footnotes_compact(writer, config)?;

        writeln!(writer)?;
        Ok(())
    }

    /// Writes this diagnostic to `stderr` using the specified [`Config`].
    #[inline]
    pub fn eprint(&self, config: &Config) -> std::io::Result<()> {
        let mut eout = BufWriter::new(std::io::stderr());
        self.write_to(&mut eout, config)?;
        Ok(())
    }

    /// Writes this diagnostic to `stderr` using the specified [`Config`]
    /// in compact mode.
    #[inline]
    pub fn eprint_compact(&self, config: &Config) -> std::io::Result<()> {
        let mut eout = BufWriter::new(std::io::stderr());
        self.write_to_compact(&mut eout, config)?;
        Ok(())
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use crate::test::diagnostic_snapshot;

    #[test]
    fn test_singleline() {
        let src = Source::new(crate::test::RUST_SAMPLE_1, Some("src/lib.rs"));
        let diagnostic =
            Diagnostic::new("error[E0072]: recursive type `List` has infinite size".red())
                .with_label(Label::new(53..66u32, ""))
                .with_label(Label::new(83..87u32, "recursive without indirection"))
                .with_footnote("error: could not compile `playground` (lib) due to previous error")
                .with_source(src);

        diagnostic.eprint(&Config::default()).unwrap();
        diagnostic_snapshot!(diagnostic);
    }

    #[test]
    fn test_multiline_1() {
        let src = Source::new(crate::test::RUST_SAMPLE_2, Some("src/main.rs"));
        let diagnostic =
            Diagnostic::new("error[E0277]: `Rc<Mutex<i32>>` cannot be sent between threads safely".red())
                .with_label(Label::new(
                    247..260u32,
                    "required by a bound introduced by this call",
                ))
                .with_label(Label::new(
                    261..357u32,
                    "`Rc<Mutex<i32>>` cannot be sent between threads safely",
                ))
                .with_footnote("help: within `{closure@src/main.rs:11:36: 11:43}`, the trait `Send` is not implemented for `Rc<Mutex<i32>>`")
                .with_footnote("note: required because it's used within this closure")
                .with_source(src);

        diagnostic.eprint(&Config::default()).unwrap();
        diagnostic_snapshot!(diagnostic);
    }

    #[test]
    fn test_multiline_2() {
        let src = Source::new(crate::test::TEXT_SAMPLE_2, Some("just testing"));
        let diagnostic = Diagnostic::new("note: this is a test".green())
            .with_label(Label::new(0..36u32, "just testing two multilines"))
            .with_label(Label::new(10..24u32, "hi"))
            .with_label(Label::styled(28u32..35u32, "hello", Style::default().red()))
            .with_source(src);

        diagnostic.eprint(&Config::default()).unwrap();
        diagnostic_snapshot!(diagnostic);
    }
}
