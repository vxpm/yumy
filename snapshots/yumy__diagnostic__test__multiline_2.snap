---
source: src/diagnostic.rs
expression: clean
---
note: this is a test
  @ [just testing]
1 │ ┬   aaaaaaa
2 │ │ ┬ bbbbbbbbbbbbbb
3 │ │ ┼ ccc
  : │ ╰─ hi
4 │ │ ┬ dddddd
5 │ ┼ ┼ eeeeeeeeee
  : ╰─┼─ just testing two multilines
  :   ╰─ hello


